File: Chapter14/R/Example12.R
# Chapter 14 - Παραμετρικοί Έλεγχοι για Δύο Πληθυσμούς και ANOVA
# Example 12: Ζευγαρωτός Έλεγχος και Σύγκριση Αναλογιών (Κλινική Μελέτη)

set.seed(101)
n_pair <- 20

# --- Paired t-test ---
before <- round(rnorm(n_pair, mean = 148, sd = 14), 1)
change <- round(rnorm(n_pair, mean = 12.5, sd = 5),  1)
after  <- before - change
d      <- before - after

cat("=== Paired t-test ===\n")
cat(sprintf("Mean before : %.2f mmHg\n", mean(before)))
cat(sprintf("Mean after  : %.2f mmHg\n", mean(after)))
cat(sprintf("Mean diff   : %.3f mmHg\n", mean(d)))
cat(sprintf("SD of diff  : %.3f\n", sd(d)))

sw <- shapiro.test(d)
cat(sprintf("\nShapiro-Wilk (differences): W = %.4f, p = %.4f\n",
            sw$statistic, sw$p.value))
cat("Normality:", ifelse(sw$p.value > 0.05, "Supported", "Violated"), "\n")

t_pair <- t.test(before, after, paired = TRUE, alternative = "greater")
print(t_pair)

# Manual calculation
d_bar   <- mean(d);  s_d <- sd(d);  se_d <- s_d / sqrt(n_pair)
t_man   <- d_bar / se_d
p_man   <- pt(t_man, df = n_pair - 1, lower.tail = FALSE)
t_crit  <- qt(0.975, df = n_pair - 1)
ci_lo   <- d_bar - t_crit * se_d
ci_hi   <- d_bar + t_crit * se_d
cohens_d_pair <- d_bar / s_d

cat(sprintf("\nManual: t = %.4f, df = %d, p (one-sided) = %.4f\n",
            t_man, n_pair - 1, p_man))
cat(sprintf("95%% CI for mu_d: [%.3f, %.3f] mmHg\n", ci_lo, ci_hi))
cat(sprintf("Cohen's d (paired): %.4f\n", cohens_d_pair))
cat("Decision:", ifelse(p_man < 0.05, "Reject H0", "Fail to reject H0"), "\n")

# --- Two-proportion z-test ---
cat("\n=== Two-Proportion Z-test ===\n")
n1_p <- 35;  x1_p <- 22
n2_p <- 30;  x2_p <- 15
p1_hat <- x1_p / n1_p;  p2_hat <- x2_p / n2_p

cat(sprintf("n1*p1 = %.1f,  n1*(1-p1) = %.1f\n",
            n1_p*p1_hat, n1_p*(1-p1_hat)))
cat(sprintf("n2*p2 = %.1f,  n2*(1-p2) = %.1f\n",
            n2_p*p2_hat, n2_p*(1-p2_hat)))

prop_res <- prop.test(c(x1_p, x2_p), c(n1_p, n2_p),
                      alternative = "two.sided", correct = FALSE)
print(prop_res)

# Manual
p_bar    <- (x1_p + x2_p) / (n1_p + n2_p)
se_prop  <- sqrt(p_bar * (1-p_bar) * (1/n1_p + 1/n2_p))
z_prop   <- (p1_hat - p2_hat) / se_prop
p_prop   <- 2 * pnorm(-abs(z_prop))
se_ci    <- sqrt(p1_hat*(1-p1_hat)/n1_p + p2_hat*(1-p2_hat)/n2_p)
ci_prop  <- c((p1_hat - p2_hat) - 1.96*se_ci,
              (p1_hat - p2_hat) + 1.96*se_ci)
h_cohen  <- 2*asin(sqrt(p1_hat)) - 2*asin(sqrt(p2_hat))
h_lbl    <- ifelse(abs(h_cohen) < 0.2, "small",
            ifelse(abs(h_cohen) < 0.5, "medium", "large"))

cat(sprintf("\nPooled p-bar: %.4f\n", p_bar))
cat(sprintf("z-statistic : %.4f,  p-value: %.4f\n", z_prop, p_prop))
cat(sprintf("95%% CI (p1-p2): [%.4f, %.4f]\n", ci_prop[1], ci_prop[2]))
cat(sprintf("Cohen's h   : %.4f (%s effect)\n", h_cohen, h_lbl))
cat("Decision:", ifelse(p_prop < 0.05, "Reject H0", "Fail to reject H0"), "\n")

# --- Visualization ---
library(ggplot2); library(gridExtra)

pair_df <- data.frame(
  Subject = rep(1:n_pair, 2),
  Value   = c(before, after),
  Time    = factor(rep(c("Before","After"), each = n_pair),
                   levels = c("Before","After"))
)
p1 <- ggplot(pair_df, aes(x = Time, y = Value, group = Subject)) +
  geom_line(colour = "steelblue", alpha = 0.45) +
  geom_point(aes(colour = Time), size = 2.5) +
  stat_summary(aes(group = 1), fun = mean, geom = "line",
               colour = "darkred", linewidth = 1.5) +
  stat_summary(aes(group = 1), fun = mean, geom = "point",
               colour = "darkred", size = 4, shape = 18) +
  labs(
    title    = "Blood Pressure: Before vs After",
    subtitle = sprintf("Paired t = %.3f, p = %.4f, d = %.3f",
                       t_man, p_man, cohens_d_pair),
    y = "Systolic BP (mmHg)", x = ""
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "none")

prop_df <- data.frame(
  Group       = c("Treatment A", "Treatment B"),
  Improved    = c(p1_hat, p2_hat),
  NotImproved = c(1-p1_hat, 1-p2_hat)
)
p2_plot <- ggplot(prop_df, aes(x = Group, y = Improved, fill = Group)) +
  geom_col(width = 0.45, alpha = 0.75) +
  geom_errorbar(
    aes(ymin = Improved - 1.96*sqrt(Improved*(1-Improved)/c(n1_p, n2_p)),
        ymax = Improved + 1.96*sqrt(Improved*(1-Improved)/c(n1_p, n2_p))),
    width = 0.12, linewidth = 0.8) +
  scale_y_continuous(labels = scales::percent, limits = c(0, 1)) +
  labs(
    title    = "Proportion Improved by Treatment",
    subtitle = sprintf("z = %.3f, p = %.3f, Cohen's h = %.3f",
                       z_prop, p_prop, h_cohen),
    y = "Proportion Improved", x = ""
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "none")

grid.arrange(p1, p2_plot, ncol = 2)

File: Chapter14/Python/Example12.py
# Chapter 14 - Παραμετρικοί Έλεγχοι για Δύο Πληθυσμούς και ANOVA
# Example 12: Ζευγαρωτός Έλεγχος και Σύγκριση Αναλογιών (Κλινική Μελέτη)

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from statsmodels.stats.proportion import proportions_ztest

np.random.seed(101)
n_pair = 20

# --- Paired t-test ---
before = np.round(np.random.normal(148, 14, n_pair), 1)
change = np.round(np.random.normal(12.5, 5, n_pair), 1)
after  = before - change
d      = before - after

print("=== Paired t-test ===")
print(f"Mean before  : {before.mean():.2f} mmHg")
print(f"Mean after   : {after.mean():.2f}  mmHg")
print(f"Mean diff    : {d.mean():.3f} mmHg")
print(f"SD of diff   : {d.std(ddof=1):.3f}")

w, p_sw = stats.shapiro(d)
print(f"\nShapiro-Wilk (differences): W = {w:.4f}, p = {p_sw:.4f}")
print("Normality:", "Supported" if p_sw > 0.05 else "Violated")

t_pair, p_pair_two = stats.ttest_rel(before, after)
p_pair_one = p_pair_two / 2 if t_pair > 0 else 1 - p_pair_two / 2
print(f"\nscipy ttest_rel: t = {t_pair:.4f}, "
      f"p (one-sided) = {p_pair_one:.4f}")

# Manual
d_bar  = d.mean();  s_d = d.std(ddof=1);  se_d = s_d / np.sqrt(n_pair)
t_man  = d_bar / se_d
p_man  = stats.t.sf(t_man, df=n_pair-1)
t_crit = stats.t.ppf(0.975, df=n_pair-1)
ci_d   = [d_bar - t_crit*se_d, d_bar + t_crit*se_d]
cohens_d_pair = d_bar / s_d

print(f"Manual: t = {t_man:.4f}, df = {n_pair-1}, "
      f"p (one-sided) = {p_man:.4f}")
print(f"95% CI for mu_d : [{ci_d[0]:.3f}, {ci_d[1]:.3f}] mmHg")
print(f"Cohen's d (paired): {cohens_d_pair:.4f}")
print("Decision:", "Reject H0" if p_man < 0.05 else "Fail to reject H0")

# --- Two-proportion z-test ---
print("\n=== Two-Proportion Z-test ===")
n1_p, x1_p = 35, 22
n2_p, x2_p = 30, 15
p1_h, p2_h = x1_p/n1_p, x2_p/n2_p

for lbl, n, p in [("A", n1_p, p1_h), ("B", n2_p, p2_h)]:
    print(f"Group {lbl}: n*p = {n*p:.1f}, n*(1-p) = {n*(1-p):.1f}")

z_sm, p_sm = proportions_ztest([x1_p, x2_p], [n1_p, n2_p],
                               alternative='two-sided')
print(f"\nstatsmodels: z = {z_sm:.4f}, p = {p_sm:.4f}")

# Manual
p_bar   = (x1_p + x2_p) / (n1_p + n2_p)
se_pool = np.sqrt(p_bar*(1-p_bar)*(1/n1_p + 1/n2_p))
z_man   = (p1_h - p2_h) / se_pool
p_man_z = 2 * stats.norm.cdf(-abs(z_man))
se_ci   = np.sqrt(p1_h*(1-p1_h)/n1_p + p2_h*(1-p2_h)/n2_p)
ci_p    = [(p1_h-p2_h) - 1.96*se_ci, (p1_h-p2_h) + 1.96*se_ci]
h_cohen = 2*np.arcsin(np.sqrt(p1_h)) - 2*np.arcsin(np.sqrt(p2_h))
h_lbl   = ("small"  if abs(h_cohen) < 0.2 else
           "medium" if abs(h_cohen) < 0.5 else "large")

print(f"Manual: z = {z_man:.4f}, p = {p_man_z:.4f}")
print(f"95% CI (p1-p2): [{ci_p[0]:.4f}, {ci_p[1]:.4f}]")
print(f"Cohen's h : {h_cohen:.4f} ({h_lbl} effect)")
print("Decision:", "Reject H0" if p_man_z < 0.05 else "Fail to reject H0")

# --- Visualization ---
fig, axes = plt.subplots(1, 2, figsize=(14, 6))

for i in range(n_pair):
    axes[0].plot(['Before','After'], [before[i], after[i]],
                 'o-', color='steelblue', alpha=0.35)
means_pair = [before.mean(), after.mean()]
axes[0].plot(['Before','After'], means_pair, 'D-',
             color='darkred', markersize=8, linewidth=2.5,
             label='Group mean')
axes[0].set_title(
    f"Blood Pressure: Before vs After\n"
    f"t = {t_man:.3f}, p = {p_man:.4f}, d = {cohens_d_pair:.3f}")
axes[0].set_ylabel("Systolic BP (mmHg)")
axes[0].legend()

grps = ['Treatment A', 'Treatment B']
props = [p1_h, p2_h]
ns    = [n1_p, n2_p]
errs  = [1.96*np.sqrt(p*(1-p)/n) for p, n in zip(props, ns)]
axes[1].bar(grps, props, color=['steelblue','cornflowerblue'],
            width=0.4, alpha=0.75)
axes[1].errorbar(grps, props, yerr=errs,
                 fmt='none', color='black', capsize=6, linewidth=1.5)
axes[1].set_ylim(0, 1)
axes[1].yaxis.set_major_formatter(
    plt.FuncFormatter(lambda v, _: f"{v:.0%}"))
axes[1].set_title(
    f"Proportion Improved by Treatment\n"
    f"z = {z_man:.3f}, p = {p_man_z:.3f}, h = {h_cohen:.3f}")
axes[1].set_ylabel("Proportion Improved")

plt.tight_layout()
plt.savefig('paired_and_proportions.png', dpi=300, bbox_inches='tight')
